### PR TITLE
Module engine version tracking

### PR DESCRIPTION
**Pull request by sipsma**
Goals:
1. Add engine version module was most recently updated with to dagger.json for the purposes of tracking as metadata
   * Currently updates on `dagger mod init/install/sync`
   * This also opens up a path to serving modules compatible schemas based on their version
     * This PR doesn't include all the internal functionality for doing so, but in the future if we make backwards incompatible changes we'll at least have the version available so we could implement this if we felt compelled.
2. Fix problem where module cache is not invalidated when engine version changes
   * This problem impacted dagger devs all the time since a change to the engine implementation often did not result in test execution being invalidated, causing us to typically have to manually clear the dev engine cache before running module tests to be sure they were actually testing our changes.
   * However, this same problem could also manifest for users if they updated their engine but ended up using local/cloud cache from older engines.
3. Publish engine builds off main to an image tag w/ their git commit rather than a global+mutable "main" tag
   * We've wanted to do this for a while since it makes it easier to just download a CLI built off main and use it with the right corresponding engine.
   * This is not *strictly* required to be here, but it felt like a good time to fix while I was in this general area
   * cc @gerhard, let me know if there's any concerns around publishing those images. We'll have a new image tag per engine commit on main, so a lot more tags than previously.
4. Remove CLI from being embedded in the engine image + in nested execs
   * This hasn't been needed for a while and may as well be removed since it's slightly increasing the size of the engine.
5. Fix problem where local imports were always retransferred in full ([details](https://github.com/dagger/dagger/pull/6469#issuecomment-1933575986))

Individual commit messages w/ more details on the above:
* [Support engine versioning by commit+local changes.](https://github.com/dagger/dagger/commit/3385f2557fce7464daf69ab24552c9aa41191db9)
* [Add engine version to ModuleConfig in dagger.json.](https://github.com/dagger/dagger/commit/f183d556550b520276b463e9cfcdd38414f05021)
* [Invalidate nested exec cache if engine version changes.](https://github.com/dagger/dagger/commit/07bf57d217e01bc906156be6bafc3a4f718e4a1b)

---

TODOs:
- [x] Do manual testing of the parts that can't be automated by setting up some release infra on my fork (i.e. test image publish off main, make sure version is correct, same for a tag, etc.)

### CODE REVIEW

**Review by jedevc — APPROVED** [review]
Approving this now, since the shape of this looks right to me (and I have the context for this right now that I might not have later), even though we need to wait for #6386 to merge.

**Comment by sipsma** [comment]
Just a note: looking at this again I'm realizing that it's *technically* a backwards compatible change, so can be safely done post v0.9.8. Will want it ASAP either way, but is okay if it's not until the next release.

**Review by gerhard — APPROVED** [review]
There is so much change here... I know, it builds on top of another commit.

I went through all the files and it's all very, very good. What a masterpiece!

Considering how much changes we already have in the upcoming patch release, I also think that we should keep this for the next patch, before we go ahead with the minor.

If you want to rebase this on top of `main` as soon as `0.9.8` is out, and all checks pass, I would be keen to have this merged as soon as possible so that we can take this through a few cycles before we cut `0.9.9` with all these changes.

Amazing job 🙌

File: core/integration/module_test.go
Comment (gerhard) [inline core/integration/module_test.go:4568]: Legit failure. I know that it would have been very difficult to spot considering the amount of change that this went through. All good!

**Comment by jedevc** [comment]
Sorry, just needed to give CI a kick, bunch of jobs from the incident earlier :D

File: .github/workflows/publish.yml
Comment (jedevc) [inline .github/workflows/publish.yml:80]: Note (but you already know this I'm sure): I hit this the other day - this workflow doesn't run in this PR, only on `main` so worth keeping an eye no `main` when this merges to make sure it actually works :scream:

**Review by jedevc — APPROVED** [review]
Since there's been some changes since this, approving again, all still looks good!

Lint failure is legitimate, but otherwise LGTM :tada: This one will genuinely be so helpful, no more purging the entire cache just to get one test to re-run :smile:

File: .github/workflows/publish.yml
Comment (sipsma) [reply .github/workflows/publish.yml:80]: This is indeed the scary part of the changes here. I'll see how much I can test before the release tomorrow (in the past I had fully replicated everything in my fork so I could actually e2e test a release). I'll update on how much progress I make there and we can make a call on whether it's worth the risk of merging before release tomorrow or postponing one more (😞, but not actually the end of the world)

**Comment by sipsma** [comment]
RE: https://github.com/dagger/dagger/pull/6469#discussion_r1482513933 I'm realizing how much manual testing + infra setup this is going to require to fully validate before the release tomorrow, which just doesn't feel worth it since we're gonna do a v0.9.10 next week anyways. So I'm going to move this to that milestone but we can merge this right after v0.9.9 is done so it doesn't sit around any more.

It's in a good place now after rebasing, only TODO left is the aforementioned manual testing, which will be my focus tomorrow post release.

**Comment by sipsma** [comment]
Also, while working on this locally I started seeing the behavior @marcosnils had reported recently of local imports always re-transferring in full. I found the one-line fix... https://github.com/dagger/dagger/pull/6469/commits/23d93469e7439fe9d284469b654123b00405db40#diff-375c41f6a94e797edee687b8fa151865e2192d8aa9088acaffa174d0de2c3578R228

I had previously seen this behavior occasionally but thought it usually worked as expected because I'd still see "reusing ref for local" in engine logs, but I'm pretty sure those logs only applied within the context of a single session. Setting the shared key to that value there meant local import mutable refs would never be used across sessions....

So we should pick up that very nice performance improvement after merging this tomorrow too 😅

File: core/integration/module_test.go
Comment (sipsma) [reply core/integration/module_test.go]: Yeah this PR has had a tumultuous history at this point 😄. Can finally bring it to a close tomorrow post-release. Fixed all the rebasing now either way.

File: engine/client/client.go
Comment (jedevc) [inline engine/client/client.go:228]: Hey, nice find! Can't believe I didn't notice this one, I've been staring at this file for hours this week :tada:

Ok, so I think this is fine - but is there any reason we couldn't make this the "client unique identifier" (the hostname?). The `sharedKey` here and the `SharedKeyHint` combine to make the `llb.Local` have an isolated space, but is there a reason we couldn't have that info in the `sharedKey` directly? (I think this is all the `sharedKey`s are used for, but I could be missing something).

Also, unrelated, but we should definitely try and have something better than the client hostname. Buildx does a cool thing by combining the directory path and a unique "node identifier", which it stores in `.config/buildx/.buildNodeID`: https://github.com/docker/buildx/blob/8f8ed68b6111552e3d4acda8ba832a735c4c1300/build/build.go#L392

**Comment by sipsma** [comment]
Okay tested the publish workflows with this PR in my fork, feel good about it. Tested:
* Engine+CLI publish off `main`
   * publishes image with git commit SHA (rather than main)
   * uploads CLIs archives; unpacking and running `dagger version` shows git commit SHA
      * e.g. `dagger aa10fe84ae0c6fd19dea8d101a8666604c79a598 (registry.dagger.io/engine) linux/arm64`
   * provision workflows pass, running some tests locally w/ `_EXPERIMENTAL_DAGGER_RUNNER_HOST=...` passes
* Engine+CLI publish off semver tag
   * publishes image with semver tag (i.e. `:v0.0.9`)
   * uploads CLIs archives, `dagger version` shows semver tag
   * I eventually hit [errors here](https://github.com/sipsma/dagger/actions/runs/7839137164/job/21391684448#step:4:4281) because I don't have nixpkgs setup or homebrew setup to publish to, but feel decent enough that the parts impacted by this PR work here

I did run into some independent problems along the way (besides the normal GHA nonsense):
* https://github.com/dagger/dagger/pull/6632 (easy fix)
* https://github.com/dagger/dagger/issues/6630
   * No clue why this was happening to me in the publish workflow yet, it seems like some obscure corner case in the shim's secrete scrubbing Trie is getting hit?
   * I included a low-risk simple fix to that problem [here](https://github.com/dagger/dagger/pull/6469/commits/1b64ceab78a5a2ec04a67462ff7cd572576f6954), though when I was hitting that panic it was in the stable engine, so won't get picked up by our runners until we actually finish the release.

I will leave it up to @gerhard and @jedevc to decide if you want to merge this before the release tomorrow or right after (I will be out so need someone else to press the merge button). I feel good about it now and it would be nice to get it out sooner than later, but up to you 🙂

File: engine/client/client.go
Comment (sipsma) [reply engine/client/client.go:234]: > The sharedKey here and the SharedKeyHint combine to make the llb.Local have an isolated space

I feel like just for simplicity purposes it would be better to only use `SharedKeyHint` on the LLB for the cache reuse. As far as I can remember the fact that they are all separate out by `:` here means we can get the same end effect by just changing `SharedKeyHint` and leaving `SharedKey` alone: https://github.com/sipsma/buildkit/blob/89bbb62a1e295391fd55fccc1f86f479de064de1/source/local/source.go#L178-L178 ?

I could be misremembering the details of how that works though. But if that's right, I just like having only one place to think about and just leaving this one empty.

> Also, unrelated, but we should definitely try and have something better than the client hostname. Buildx does a cool thing by combining the directory path and a unique "node identifier", which it stores in .config/buildx/.buildNodeID

Huh yeah, I have some built-in aversion to creating files in our user's homedirs, but it's really not that big of a deal and could be useful here I agree. Not sure how beneficial this ends up being in practice since it would only be useful for laptop/desktop use cases, as opposed to "ephemeral CI runner" cases. And I think hostname is a pretty reasonable heuristic for identity among the laptop/desktop cases, but definitely not flawless. Either way, no opposed, especially if we can re-use code and get it for mostly free 👍

**Review by jedevc — APPROVED** [review]

**Pull request merged**
